### ISSUE TITLE
Tech discovery: Investigate creating mail task from VBMS document types

### ISSUE THREAD

**Issue by laurjpeterson** [department-of-veterans-affairs/caseflow#6763]
Investigate the feasibility of querying Veteran eFolders to identify folders with new documents, and creating mail tasks based off of those new documents. 

Potential flow:
- Caseflow recognizes Veterans with open or closed appeals 
- Caseflow recognizes VBMS documents of certain types - e.g. 
- Caseflow adds a mail task o

Questions:
- Can we use the VBMS endpoint about fetching new documents without putting too much load on VBMS? How much is too much load?

### Resources
- Surfacing actionable mail epic - https://github.com/department-of-veterans-affairs/caseflow/issues/5950

Nicholas went through and mapped VACOLS mail types to VBMS document types - see below. This list of VBMS document types is the only type of document we should detect (pending changing requirements). 

> This list is limited to actionable mail. Other mail types may of course come through the mail portal, and would need to have appropriate labels, as they do today. The thought here is to have these programatically identified by Caseflow when added to an active eFolder.

> Uncertainties are italicized.

VACOLS field | VBMS Category | VBMS Document Type | Is Nicholas' confident in this mapping?
---|---|---|---|
01	Change of Address | 1. Correspondence<br>2. Correspondence | 1. Change of Address/Direct Deposit Request<br>2. VA 572 Request for Change of Address / Cancellation of Direct Deposit | Yes
02	Congressional Interest | Correspondence | Congressionals | Yes
~~03	Controlled Correspondence~~ | | | | Yes
04	CUE Related | Appeals | CUE Related | Yes
05	Evidence or Argument | 1. Correspondence<br>2. Correspondence<br>3. Appeals<br>4. Appeals<br>5. Correspondence<br>6. Appeals<br>7. Appeals<br>8. Correspondence<br>9. Correspondence<br>10. Appeals<br>11. Correspondence<br>12. Appeals<br>13. Appeals<br>14. Correspondence | 1. Correspondence<br>2. VA 21-4138 Statement In Support of Claim<br>3. VA 646   Statement of Accredited Representative in Appealed Case<br>4. VA 9 Appeal to Board of Appeals<br>5. Email Correspondence<br>6. Appeal Process Request Letter<br>7. Appeals Election Letter<br>8. Third Party Correspondence<br>9. Due Process Waiver<br>10. Appellate Brief (VSO IHP; Post remand Brief; Attorney Brief)<br>11. BVA-General<br>12. RAMP Opt-in Election<br>13. Notice of Disagreement<br>14. Buddy / Lay Statement | Mostly
06	FOIA Request | Correspondence | Freedom of information Act (FOIA) / Privacy Act Request | Yes
07	Hearing Related | Appeals | Hearing Request | Yes
08	Motion to Advance on Docket | Appeals | Motion to Advance on Docket | Yes
09	Motion for Reconsideration | Appeals | Motion for Reconsideration | Yes
10	Power of Attorney Related | 1. Representation<br>2. Representation | 1. Power of Attorney (incl. VA 21-22, VA 22a)<br>2. VA 21-22 Appointment of Veterans Serv. Org. as Claimant Rep | Yes
~~11	Professional Service Mail~~ |  |  | Yes
12	Returned or Undeliverable Mail | Correspondence | Returned Mail | Yes
*13	Status Inquiry* | *???* | *??? Suggest Appeals - Appeals Status Inquiry* | No
14	Death Certificate | 1. Correspondence<br>2. Non-Medical Evidence – Death Evidence | 1. VA 27-0820a Report of First Notice of Death<br>2. Death Certificate | Yes
~~15	Appellate Group Mail~~
16	Extension Request | Appeals | Extension Request | Yes
~~17	Privacy Act Request~~ | | | Yes
*18	Other Motion* | *Correspondence* | *BVA-General* | No
*19	Privacy Complaints* | *???* | *???* | No
~~20	Intra/Inter Agency Request~~ |  |  | Yes
~~21	Ebert Temporary Transfer~~ | | | Yes
~~22	How Do I Appeal Pamphlet~~ | | | Yes
~~23	ECA Revocation~~ |  |  | Yes
24	FOIA Request & Other Actions | Correspondence | FOIA/Privacy Act Request | Yes
*25	Motion to Vacate* | *Correspondence* | *BVA-General (alt suggest Appeals - Motion to Vacate* | No
*26	Attorney Inquiry* | *???* | *??? (see Status Inquiry)* | No
*27	Withdrawal of Appeal* | *???* | *Suggest Appeals - Appeal Withdrawal* | No

**Comment by mdbenjam** [comment]
# Tech spec

![image](https://user-images.githubusercontent.com/3885236/44925872-9d543b00-ad1d-11e8-8b22-5345782be9be.png)
^ It's mail time

# Overview
We are currently getting a list of documents from VBMS for every appeal in a user's Queue whenever they view their Queue. The new document tab is shown to users based on comparing the list of documents, and the time each of those documents was received, to the time the user last viewed the documents in Reader. This answers the question which documents are new to this user.

The description above is asking a different question, what documents are new in VBMS. This question is harder to answer since it's unbounded. Currently the number of queries we make to VBMS is related to the number of board employees (n), while getting information on any new mail is related to the number of Veterans who have a appeals (m). Querying for every appeal we're aware of to determine if there are new documents would be way to intensive, and would likely put an undue stress on our dependencies.

Luckily, according to the documentation, there is a VBMS endpoint called `findVeteransWithFolderModifications`.

> findVeteransWithFolderModifications searches for Veterans that have had at least one document uploaded matching the TypeCategory information within the supplied DateTimeRange. For safety, the service will return a maximum of 5000 results. If the corresponding error message is received for the scenario where the maximum has been exceeded, please narrow your DateTimeRange and try again.

# Solution

**Comment by laurjpeterson** [comment]
The alternative to this approach is to ask a BVA employee to manually create a mail task, just like the Mail team currently creates a mail tab in VACOLS today.

**Comment by nicholasholtz** [comment]
@laurjpeterson In the alternative scenario, what sort of development lift is necessary to permit that sort of task creation? Is this a scenario where a manual mail processing could occur to begin with, with task creation, and the heavier lift of automation occurs later, post-October 1 sometime?

@mdbenjam You indicate a month's worth of work - is that one engineer working alone?

**Comment by mdbenjam** [comment]
@nicholasholtz I would budget one month for me to do it.

**Comment by laurjpeterson** [comment]
@nicholasholtz - the alternatives are:

Option 1: Create a Caseflow task
- mail employee clicks upload in VBMS (note, sometimes takes 24 hours for document to show in VBMS)
- mail employee searches for Veteran in Caseflow - *development complete*
- mail employee clicks button (or something) to add a mail task - *would need to build*
- mail employee enters in information about the mail task - *would need to build*
   - mail type (pre-determined or free text)
   - mail source?
- mail employee choose who to assign mail to (including only users who have queues) - *would need to build*
   - if want to assign to a user who doesn't have a queue, mail employee has to email the mail task

Option 2: Email mail tasks
- mail employee clicks upload in VBMS (note, sometimes takes 24 hours for document to show in VBMS)
- mail employee emails Board user who needs to process the mail information about the mail task

**Comment by amprokop** [comment]
This tech spec looks great. A few thoughts:

- How sure are we that we can fully capture the Mail team's activities through a mapping of VBMS document type to some type of action? If we were later to find out that there are some or many ambiguous edge cases that need manual user intervention, it surely would not be the first time.

- If we are not fully sure that we can automatically and correctly categorize all pieces of mail and create actionable tasks, what will we do if mail appears that we can't automatically categorize? Would we essentially have to build the manual Caseflow mail task creation piece as well?

- Automating mail task creation doesn't appear to be critical for AMA implementation, unless I'm missing something.

- Would building a manual mail task creation piece up front help us de-risk the business end by checking the correctness of an automated mail task creation system later on? For example, if users were checking the mail portal and creating `MailTasks`, theoretically we could build this automated piece which would create `FakeMailTasks`, then check to see how the results are different. Potentially if we had time we could also use VACOLS mail to accomplish the same end.

**Comment by laurjpeterson** [comment]
Thanks @amprokop, all good thoughts. More about the idea itself and how it solves (or doesn't completely solve) the business problem than the technical feasibility of it, so I'll weigh in.

> How sure are we that we can fully capture the Mail team's activities through a mapping of VBMS document type to some type of action? If we were later to find out that there are some or many ambiguous edge cases this team handles that need manual user intervention, it surely would not be the first time.
- We've gotten this indication from the Mail team's SOP and routing sheet. They look at the VBMS document and create a mail task based on the document type or based on where the appeal is at the Board. 
   - [routing rules - October 2017](https://github.com/department-of-veterans-affairs/appeals-design-research/blob/master/Projects/Caseflow%20Queue/Research/Documents/Mail%20Assignment%20for%20October%202017.pdf)
   - [SOP](https://github.com/department-of-veterans-affairs/appeals-design-research/blob/master/Projects/Caseflow%20Queue/Research/Documents/PrinciplesOfMailManagementOct2016.pdf)
- We can potentially restrict the number of VBMS document types to just those @nicholasholtz pointed out in his mapping of VACOLS mail types to VBMS document types, since we can't easily differentiate VBA PO box from BVA PO box within VBMS right now

> If we are not fully sure that we can automatically and correctly categorize all pieces of mail and create actionable tasks, what will we do if mail appears that we can't automatically categorize? Would we essentially have to build the manual Caseflow mail task creation piece as well?
- This mail would be assigned based on where the appeal is - based on the most recent task added to the appeal. For example, if the appeal is at the attorney, and we don't know what the mail is about, we'd still assign the case to VLJ management / co-located branch just like we would if we knew the mail was specifically a POA clarification

> Automating mail task creation doesn't appear to be critical for AMA implementation, unless I'm missing something.
- This is a good callout point. We do need to support processing mail for AMA appeals, but the level of support and automation is open to interpretation. 
- What we do need to do, though, is explain to the mail team if they will or won't be using Caseflow to process mail in the short term, for the upcoming October RAMP appeals pilot. 

> Would building a manual mail task creation piece up front help us de-risk the business end by checking the correctness of an automated mail task creation system later on? For example, if users were checking the mail portal and creating MailTasks, theoretically we could build this automated piece which would create FakeMailTasks, then check to see how the results are different. Potentially if we had time we could also use VACOLS mail to accomplish the same end.
- These are great ideas, thanks for listing them out. I'd be open to de-risking the approach. Building the creation of a generic task seems inevitable anyway - specifically thinking about adding a task for a post-decisional motion, or something similar.

#### Path forward
- Based on the risk in this approach, and a few comments on this thread, it sounds like pursuing the automated option in the short term is not ideal
- @allyceh and @sneha-pai are already working on designs for users creating a mail task in Caseflow and viewing that mail task in a queue
- @allyceh, @sneha-pai, and I are meeting with LRP to discuss the two options outlined above so we know how the mail team should work in October. There are still unknowns about how long Option 1 would take to build, but I'd assume that would be their preferred approach

I'm still motivated to get creative around how to speed up mail processing for Veterans, especially mail that could have an impact on their case (e.g. AOD motions) or that unblocks Board teams from doing their work (e.g. co-located acting on the POA clarification mail without delay, so the attorney can get back to writing a decision). This is not something I have a baseline of data or turnaround times for because VACOLS data on tracking mail, abeyance, etc. is messy. But we can postpone implementation after some data profiling/testing in order to be sure this is right.

**Comment by amprokop** [comment]
Potentially we should only ask them about the email option- I’m not super
confident we will have bandwidth to handle building a Mail solution before
October.
